packages/task/src/browser/task-service.ts

```diff
@@ -0,0 +1,158 @@
+            if (event.ctx !== this.getContext()) {
+                return;
+            }
+
+            if (event.code === 0) { /* normal process exit */
+                let success = '';
+
+                // this finer breakdown will not work on Windows.
+                if (signal && signal !== '0') {
+                    if (signal === '1') {
+                        success = 'Terminal Hangup received - ';
+                    } else if (signal === '2') {
+                        success = 'User Interrupt received - ';
+                    } else if (signal === '15' || signal === 'SIGTERM') {
+                        success = 'Termination Interrupt received - ';
+                    } else {
+                        success = 'Interrupt received - ';
+                    }
+                } else {
+                    success = 'Success - ';
+                }
+
+                success += `Task ${event.taskId} has finished. exit code: ${event.code}, signal: ${event.signal}`;
+                this.messageService.info(success);
+            } else { /* abnormal process exit */
+                this.messageService.error(`Error: Task ${event.taskId} failed. Exit code: ${event.code}, signal: ${event.signal}`);
+            }
+        });
+    }
+
+    public async getTasks(): Promise<string[]> {
+        return Promise.resolve(this.taskConfigurations.getTaskLabels());
+    }
+
+    public getRunningTasks(): Promise<TaskInfo[]> {
+        return this.taskServer.getRunningTasks(this.getContext());
+    }
+
+    public async run(taskName: string): Promise<void> {
+        let taskInfo;
+        try {
+            const task = await this.taskConfigurations.getTask(taskName);
+            taskInfo = await this.taskServer.run(this.massageTaskConfiguration(task), this.getContext());
+        } catch (error) {
+            this.logger.error(`Error launching task '${taskName}': ${error}`);
+            this.messageService.error(`Error launching task '${taskName}': ${error}`);
+            return;
+        }
+
+        this.logger.debug(`Task created. task id: ${taskInfo.taskId}, OS ProcessId: ${taskInfo.osProcessId} `);
+
+        // open terminal widget if the task is based on a terminal process:
+        if (taskInfo.terminalId !== undefined) {
+            this.attachTerminal(taskInfo.terminalId, taskInfo.taskId);
+        }
+    }
+
+    public async attach(terminalId: number, taskId: number): Promise<void> {
+        this.attachTerminal(terminalId, taskId);
+    }
+
+    // Perform some adjustments to the task launch configuration:
+    // We accept some symbolic value in a task configuration. ATM we support
+    // using "$workspace" as a symbolic value as part of parameter "cwd". Here
+    // we replace it by its value, for the current client
+    // also, if launch config 'cwd' is not defined, use WS Root as default.
+    protected massageTaskConfiguration(task: TaskOptions): TaskOptions {
```

Comment:
You should use proper js doc (ML comments), so the tools can pick it up.

Reply:
done